### PR TITLE
Make InferenceValidator inherit from BaseTypeValidator

### PR DESCRIPTION
**Pull request by topnessman**
Return real top annotation for real AnnotationMirror in
InferenceQualifierHierarchy#getTopAnnotationInHierarchy()

### CODE REVIEW

File: src/checkers/inference/InferenceQualifierHierarchy.java [hunk 1]
```diff
@@ -325,7 +325,7 @@ public AnnotationMirror getTopAnnotation(final AnnotationMirror am) {
             return varAnnot;
         } // else
 
-        return this.getTopAnnotations().iterator().next();
+        return inferenceMain.getRealTypeFactory().getQualifierHierarchy().getTopAnnotations().iterator().next();
```
Comment (jyluo) [inline src/checkers/inference/InferenceQualifierHierarchy.java]: There's a TODO comment in line 319 above that needs an update.

**Comment by jyluo** [comment]
There seems to be two different changes in this PR, one for InferenceValidator inheritance and another for InferenceQualifierHierarchy#getTopAnnotationInHierarchy(). Are they related?

**Comment by topnessman** [comment]
Not exactly, but I remember it solves one problem when I ran PICOInfer on real project.

File: src/checkers/inference/InferenceQualifierHierarchy.java (hunk 1 above)
Comment (wmdietl) [reply src/checkers/inference/InferenceQualifierHierarchy.java]: This changes makes getTopAnnotation and getBottomAnnotation consistent, which seems like a good thing.
I'm not sure what update to that TODO comment would improve the situation.

File: src/checkers/inference/InferenceValidator.java [hunk 2]
```diff
@@ -1,402 +1,24 @@
...
  * A visitor to validate the types in a tree.
  */
-public class InferenceValidator extends AnnotatedTypeScanner<Void, Tree> implements TypeValidator {
-    protected boolean isValid = true;
+public class InferenceValidator extends BaseTypeValidator {
 
-    protected final BaseTypeChecker checker;
-    protected final InferenceVisitor<?, ?> visitor;
-    protected final AnnotatedTypeFactory atypeFactory;
+    protected boolean infer;
```
Comment (wmdietl) [inline src/checkers/inference/InferenceValidator.java]: What is the point of this boolean? Nobody reads its value and setting it doesn't have an effect. Can this be removed?

File: src/checkers/inference/InferenceQualifierHierarchy.java (hunk 1 above)
Comment (CharlesZ-Chen) [inline src/checkers/inference/InferenceQualifierHierarchy.java]: Raise error to see if we still have the case of requiring real qual from InferQualHier.

File: src/checkers/inference/InferenceValidator.java (hunk 2 above)
Comment (CharlesZ-Chen) [inline src/checkers/inference/InferenceValidator.java]: Update javadoc. Explain the purpose of this field.

Think about how to share this flag.

**Comment by topnessman** [comment]
Ready to be reviewed

**Pull request merged**